arm: board: Add support for board Atmel SAM4S Xplained.

fallrisk opened the pull request:
This commit provides support for Atmel SAM4S Xplained board
Pinmux driver is provided with initial support definitions.
External SRAM support is also provided.

Signed-off-by: fallrisk <jwatson5@gmail.com>

fallrisk committed:
arm: board: Add support for board Atmel SAM4S Xplained.

This commit provides support for Atmel SAM4S Xplained board

Signed-off-by: Justin Watson <jwatson5@gmail.com>

galak reviewed (changes requested):
Need board doc

dbkinder commented on boards/arm/sam4s_xplained/doc/sam4s_xplained.rst:
```diff
@@ -0,0 +1,159 @@
+===================
+
+The `SAM4S Xplained Online User Guide`_ has detailed information about board
+connections. You can find a PDF of the schematic at http://www.atmel.com/tools/sam4s-xpld.aspx?tab=documents.
```
unless you REALLY want this url to be displayed in the doc, it would be better to give this URL a title and put the title here instead (as you did for the `SAM4S Xplained Online User Guide` above.

dbkinder commented on boards/arm/sam4s_xplained/doc/sam4s_xplained.rst:
```diff
@@ -0,0 +1,159 @@
+Flashing
+========
+
+#. Build the Zephyr kernel and the application:
```
how about 

    Build the Zephyr kernel and the :ref:`hello_world` sample application:

dbkinder commented on boards/arm/sam4s_xplained/doc/sam4s_xplained.rst:
```diff
@@ -0,0 +1,159 @@
+      ROMTbl 0 [3]: FFF01000, CID: B105E00D, PID: 003BB001 ITM
+      ROMTbl 0 [4]: FFF41000, CID: B105900D, PID: 000BB9A1 TPIU
+      Found 1 JTAG device, Total IRLen = 4:
+       #0 Id: 0x4BA00477, IRLen: 04, IRPrint: 0x1, CoreSight JTAG-DP (ARM)
```
remove the extra space before the #0 to keep the indent the same

dbkinder commented on boards/arm/sam4s_xplained/doc/sam4s_xplained.rst:
```diff
@@ -0,0 +1,159 @@
+      JLink> r
+      JLink> go
+
+   You should see "Hello World!" in your terminal.
```
I think it will write out "Hello Word! arm", yes?

dbkinder reviewed (changes requested):
a few doc suggestions...

mnkp commented on dts/arm/sam4s_xplained.dts:
```diff
@@ -0,0 +1,18 @@
+
+/ {
+	model = "Atmel SAM4S Xplained Board with an Atmel SAM4S16C SoC";
+	compatible = "atmel,xplained", "atmel,sam4s16c", "atmel,sam4s";
```
The first string should be: "atmel,sam4s_xplained"

mnkp commented on boards/arm/sam4s_xplained/sam4s_xplained_defconfig:
```diff
@@ -0,0 +1,8 @@
+CONFIG_BOARD_SAM4S_XPLAINED=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_CONSOLE=y
+CONFIG_PINMUX=y
```
That's probably not the Zephyr policy but diff works much nicer if compared files have newline at the end of file.

mnkp commented on boards/arm/sam4s_xplained/ext_ram.c:
```diff
@@ -0,0 +1,122 @@
+/*
```
SMC driver should better be located somewhere in drivers/ folder as it is SoC specific, not board specific.

mnkp commented on boards/arm/sam4s_xplained/ext_ram.c:
```diff
@@ -0,0 +1,122 @@
+#include <sys_io.h>
+
+/* Static Memory Controller (SMC) */
+struct __smc_cs {
```
SMC driver should better be located somewhere in drivers/ folder as it is SoC specific, not board specific.

mnkp commented on boards/arm/sam4s_xplained/ext_ram.c:
```diff
@@ -0,0 +1,122 @@
+#include <soc.h>
+#include <sys_io.h>
+
+/* Static Memory Controller (SMC) */
```
SMC driver should better be located somewhere in drivers/ folder as it is SoC specific, not board specific.

fallrisk replied:
Where at? I don't remember seeing any external SRAM or external memory drivers.

mnkp replied:
Yes, currently there is none. The arch/arm/soc/atmel_sam/sam4s location is also worth considering. I guess it depends if we need any Kconfig options associated with the SMC driver. Maybe one of the maintainers will have some good tips on the naming/location.

fallrisk replied:
Yeah, I agree with everything you said. I didn't know where to put it. I put it here because the chip is on the board. I'm hoping Kumar can tell me what to do here.

mnkp replied:
Maybe it would be better to add support for SMC and the external IS66WV51216DALL SRAM in a separate patch. It's quite a job. One more remark: shouldn't we also update the linker script to let the compiler know we have more RAM?

fallrisk replied:
Yeah, I can take it out of the patch. I was thinking that after I responded to you earlier. There is no precedent currently for external RAM. I did this for a project I'm working on. I can just keep it in my personal branch.

mnkp replied:
It wouldn't be all that bad to have support for an external RAM. But being the first one is always tricky.

galak replied:
Seems like this is pretty board / memory connected specific.  We could have a helper function that lives in arch/arm/soc/atmel_sam/sam4s/ to do the pin setup & anything generic, but I don't see this being a driver as its seems so trivial & board/memory specific.

mnkp replied:
The way the code is written is indeed pretty board / memory specific. No offense to anyone but it feels like a quick hack. I believe we should have a proper implementation with a standalone SMC driver. The SMC module is pretty complex, apart from SRAM it can support PSRAM, PROM, EPROM, EEPROM, LCD Module, NOR Flash and NAND Flash. For now the driver code can be very simple, but if it's a standalone code we will have a place where it can grow when the need arises.

I still think it would be better to add external RAM support in a separate patch. It seems like the topic needs some more thought.

mnkp commented on boards/arm/sam4s_xplained/Kconfig:
```diff
@@ -0,0 +1,13 @@
+	depends on BOARD_SAM4S_XPLAINED
+	help
+	  This is for the IS66WV51216DALL chip on the
+	  SAM4S Xplained Board.
```
No newline at the end of file.

mnkp commented on boards/arm/sam4s_xplained/Kconfig.board:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config BOARD_ATMEL_SAM4S_XPLD
```
Should probably be: BOARD_SAM4S_XPLAINED

mnkp commented on boards/arm/sam4s_xplained/Kconfig:
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config ENABLE_EXTERAL_SRAM
```
The option name should be SAM4S specific.

galak commented on boards/arm/sam4s_xplained/ext_ram.c:
```diff
@@ -0,0 +1,122 @@
+	 * Using a local pointer because it allows us to easily change the
+	 * cs in use for the SRAM.
+	 */
+
```
remove blank line

galak commented on boards/arm/sam4s_xplained/ext_ram.c:
```diff
@@ -0,0 +1,122 @@
+	volatile struct __smc_cs *smc_cs = &__SMC->cs1;
+
+	/* Remove the pins from the PIO controller. */
+
```
remove blank line

galak commented on boards/arm/sam4s_xplained/ext_ram.c:
```diff
@@ -0,0 +1,122 @@
+
+	/* Remove the pins from the PIO controller. */
+
+	__PIOC->pdr = 0xffffffff;
```
can we clean up magic numbers?

galak commented on boards/arm/sam4s_xplained/ext_ram.c:
```diff
@@ -0,0 +1,122 @@
+	__PIOA->pdr = 0x03dc0003;
+
+	/* Set pins to their SMC peripheral function (Page 444, 26.4.1). */
+
```
remove blank line

galak commented on boards/arm/sam4s_xplained/ext_ram.c:
```diff
@@ -0,0 +1,122 @@
+	__PIOC->abcdsr2 = 0;
+
+	/* PA 0,1,18,19,20,22,23,24,25 */
+
```
remove blank line

galak commented on boards/arm/sam4s_xplained/ext_ram.c:
```diff
@@ -0,0 +1,122 @@
+
+	/* PA 0,1,18,19,20,22,23,24,25 */
+
+	__PIOA->abcdsr1 &= ~(0x03dc0003);
```
can we cleanup magic numbers?

galak commented on boards/arm/sam4s_xplained/ext_ram.c:
```diff
@@ -0,0 +1,122 @@
+	__PIOA->abcdsr2 |= 0x03dc0003;
+
+	/* Turn on the clock for the SMC. */
+
```
remove blank line

galak commented on boards/arm/sam4s_xplained/ext_ram.c:
```diff
@@ -0,0 +1,122 @@
+	__PMC->pcer0 = BIT(10);
+
+	/* Set the signal durations. */
+
```
remove blank line

galak commented on boards/arm/sam4s_xplained/ext_ram.c:
```diff
@@ -0,0 +1,122 @@
+	 * SMC CS2 0x62000000 to 0x0x63000000
+	 * SMC CS3 0x64000000 to 0x0x64000000
+	 */
+
```
are we missing setup up the actual CS0..CS3 registers?

fallrisk replied:
The external SRAM chip is connected to CS1. Only CS1 registers have been setup. I put a comment at the bottom of what the ranges were in the documentation. It could probably go in another place. I see why you got confused. Usually a comment precedes some code. I will put this comment somewhere else.

mnkp commented on boards/arm/sam4s_xplained/Makefile:
```diff
@@ -0,0 +1,12 @@
+asflags-y := ${ccflags-y}
+
+obj-$(CONFIG_PINMUX) += pinmux.o
+obj-$(CONFIG_ENABLE_EXTERAL_SRAM) += ext_ram.o
```
This line is no longer necessary, should be removed.

mnkp commented on dts/arm/Makefile:
```diff
@@ -1,6 +1,6 @@
 ifeq ($(CONFIG_HAS_DTS),y)
 dtb-$(CONFIG_BOARD_DISCO_L475_IOT1)	=	disco_l475_iot1.dts_compiled
-dtb-$(CONFIG_BOARD_FRDM_K64F)	=	frdm_k64f.dts_compiled
+dtb-$(CONFIG_BOARD_FRDM_K64F)		=	frdm_k64f.dts_compiled
```
Tempting, but we probably shouldn't do it.

pfalcon replied:
> +dtb-$(CONFIG_BOARD_FRDM_K64F)

Yep, that's the golden rule of making changes - avoid any unrelated ones, the only outcome of these can be increased review time, because reviewers will need to exert more effort to filter important changes from noise. Some smart reviewers may even put such reviews at the end of their queue ;-). Changes like that should go into separate PRs, ideally merged in minutes (though we aren't yet there).

mnkp commented on boards/arm/sam4s_xplained/doc/sam4s_xplained.rst:
```diff
@@ -0,0 +1,162 @@
+   - Parity: None
+   - Stop bits: 1
+
+#. Download JLink from the Segger `JLink Downloads Page`_. Go to the section
```
I should have commented earlier: in Zephyr we tend to use OpenOCD or other open source tools where possible rather than a proprietary ones. SAM4S Xplained board seems to be supported by OpenOCD so we probably should use it instead of JLink from Segger. In the future it would be also nice to have support for 'make flash' target. It's quite convenient and helps a lot for new users. That's said, so far no one else complained about this point and I'm also fine if this file is merged as it is at the moment.

fallrisk replied:
The board comes with JLink OB for free. I will add OpenOCD later. I have never used OpenOCD.

mnkp commented on arch/arm/soc/atmel_sam/sam4s/Kconfig.soc:
```diff
@@ -112,7 +105,15 @@ config SOC_ATMEL_SAM4S_PLLB_DIVB
+	With default of MULB == 3, and DIVB == 1,
+	PLL is running at 4 times of main clock.
+
+config SOC_ATMEL_SAM4S_EFC0_FMR_FWS
```
FWS value can be either calculated statically from SYS_CLOCK_HW_CYCLES_PER_SEC or dynamically at run time in atmel_sam4s_init() function. We should not add a new config option.

fallrisk replied:
The table 44-73 shows different values based on VDDCORE and VDDIO though.

mnkp replied:
Fair point. I believe however that in the future we will likely want to support changing processor clock frequency on the fly, so we would rather need to know VDDIO/VDDCORE values. Providing FWS as a config option is not the best approach here. Anyway, that's why it was left as a TODO, we need to figure out how to do it best.

fallrisk replied:
I don't see why it's bad to make it a configure option. A person would design a board and look at that table and choose the frequency they want to run at and appropriate VDDCORE and VDDIO. Then they would set the FWS accordingly to work with their board. It's like any other option. If we are going to change processor clock frequency on the fly then why do we even have processor speed as an option? I see this config option as the same thing as setting your PLL multiplier or divider. @galak or @nashif could you help us with this?

mnkp replied:
If we decide that we need to ask user to provide VDDCORE and VDDIO to support run time frequency modifications then this option becomes redundant. If we decide that the only way to support run time frequency modifications is to have user explicitly provide FWS value then this option would be fine. I would rather leave this discussion for some later time.

fallrisk replied:
Okay, so would it be acceptable to remove the config option and just have it fixed to 5, which is 120 MHz, VDDCORE 1.2V, VDDIO 2.7-3.6 V? Those are the values for the SAM4S Xplained.

mnkp replied:
Yes, I believe that's the best move for now. Also, setting FWS to 5 is safe. Any other board with any other supply voltage will run, just not in an optimal way.

mnkp commented on boards/arm/sam4s_xplained/sam4s_xplained_defconfig:
```diff
@@ -0,0 +1,12 @@
+CONFIG_SERIAL=y
+CONFIG_UART_SAM=y
+CONFIG_UART_SAM_PORT_1=y
+CONFIG_BOARD_SAM4S_XPLAINED=y
```
Add: CONFIG_SOC_ATMEL_SAM4S_EXT_MAINCK=y

mnkp commented on boards/arm/sam4s_xplained/pinmux.c:
```diff
@@ -0,0 +1,154 @@
+/*
```
Since SAM4S is using now ASF and a new setup based on SAME70 this file is no longer necessary. The required pin configuration is provided by soc_pinmap.h file. pinmux.c can be deleted.

mnkp reviewed (changes requested):
This looks much better, however there are still a few things to change. 

I propose to split this PR into two commits. One to fix all the issues in the core SoC, i.e. all the files in arch/arm/soc/atmel_sam/sam4s folder and the other commit that actually adds SAM4S Xplained board support.

mnkp commented on arch/arm/soc/atmel_sam/sam4s/soc.c:
```diff
@@ -184,7 +172,9 @@ static ALWAYS_INLINE void clock_init(void)
 
-	/* Setup PLLB. */
+	/*
+	 * Setup PLLB.
```
PLLB is used by USB only. Until we provide an USB driver we should not touch this PLL. It only increases power consumption.

fallrisk replied:
According to figure 29-1 (page 516) of the datasheet, it can be used by the Master Clock, the Programmable Clock (an output clock), and the USB Clock. However, making the configuration available would require more Kconfig options. I am going to leave it out for now because it would require more Kconfig options. It was in there because on my project branch I have USB.

mnkp replied:
> According to figure 29-1 (page 516) of the datasheet, it can be used by the Master Clock, the Programmable Clock (an output clock), and the USB Clock.

Good point.

mnkp commented on boards/arm/sam4s_xplained/Makefile:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+ccflags-y += -I$(srctree)/include/drivers
```
These flags definitions are no longer required, should be removed.

fallrisk replied:
Ok, I will remove. This means that the SAM3X also needs them removed because I looked at it originally when doing this work.

mnkp replied:
Not just yet. In case of SAM3X we still compile pinmux.c to support old device drivers.

fallrisk replied:
Do you mean all of these are unnecessary? If I am to remove them all should I add the statement:
 
```
obj- += dummy.o
```

mnkp replied:
Yes, lines 7-9 can be removed. And yes, we need `obj- += dummy.o`

mnkp commented:
Thanks for the update. This PR can also include all the files in arch/arm/soc/atmel_sam/sam4s folder which need to be changed. Only as a separate commit. I.e. you can submit this PR with two commits. Let me know if you have any problems.

fallrisk commented:
Hey @mnkp, I have made the code in the soc directory into a new PR. @galak he said he preferred that I split the PR. I am going to post the SoC PR so you can have a look at it. When I get home I will make a branch to combine the PRs and test it again. I tested the code last night and it worked with the UART SAM. However, since I added CONFIG_SOC_ATMEL_SAM4S_EXT_MAINCK=y I need to test it again.

mnkp reviewed (approved):
Looks good to me.

fallrisk commented:
@mnkp, @galak I added an image for the SAM4S Xplained docs.

galak commented:
@dbkinder any update on doc review?

dbkinder reviewed (approved):
LGTM

galak committed:
Merge branch 'arm' into sam4s_xplained_0
